runtime/jvmti/jvmtiHelpers.c

```diff
@@ -95,56 +95,108 @@ static UDATA watchedClassEqual (void *lhsEntry, void *rhsEntry, void *userData);
 		if (currentThread->threadObject == threadObject) {
 			*vmThreadPtr = currentThread;
 			return JVMTI_ERROR_NONE;
 		}
 	}
 
-	/* Make sure the vmThread stays alive while it is being used */
-
+	/* Make sure the vmThread stays alive while it is being used. */
 	omrthread_monitor_enter(vm->vmThreadListMutex);
-	if (!J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) || ((targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject)) == NULL)) {
+#if JAVA_SPEC_VERSION >= 19
+	isVirtualThread = IS_VIRTUAL_THREAD(currentThread, threadObject);
+	if (isVirtualThread) {
+		/* Make sure the virtual thread stays alive while it is being used. */
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		jint vthreadState = J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject);
+		j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(currentThread, threadObject);
+		if (NULL != carrierThread) {
+			targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
+		}
+		isThreadAlive = (vthreadState != JVMTI_VTHREAD_STATE_NEW) && (vthreadState != JVMTI_VTHREAD_STATE_TERMINATED);
+	} else
+#endif /* JAVA_SPEC_VERSION >= 19 */
+	{
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+		isThreadAlive = J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) && (NULL != targetThread);
+	}
+
+	if (!isThreadAlive) {
 		if (mustBeAlive) {
+#if JAVA_SPEC_VERSION >= 19
+			if (isVirtualThread) {
+				omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
+			}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 			omrthread_monitor_exit(vm->vmThreadListMutex);
 			return JVMTI_ERROR_THREAD_NOT_ALIVE;
 		}
 	}
 
 	*vmThreadPtr = targetThread;
-	if (targetThread != NULL) {
-		++(targetThread->inspectorCount);
+	if (NULL != targetThread) {
+		targetThread->inspectorCount += 1;
+	}
+#if JAVA_SPEC_VERSION >= 19
+	if (isVirtualThread) {
+		U_64 vthreadInspectorCount = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->virtualThreadInspectorCountOffset) + 1;
+		Assert_JVMTI_true(vthreadInspectorCount > 0);
+		J9OBJECT_U64_STORE(currentThread, threadObject, vm->virtualThreadInspectorCountOffset, vthreadInspectorCount);
+		omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
 	}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 	omrthread_monitor_exit(vm->vmThreadListMutex);
 	return JVMTI_ERROR_NONE;
 }
 
 
 
 void
-releaseVMThread(J9VMThread * currentThread, J9VMThread * targetThread)
+releaseVMThread(J9VMThread *currentThread, J9VMThread *targetThread, jthread thread)
 {
-	if ((currentThread != targetThread) && (targetThread != NULL)) {
-		J9JavaVM * vm = targetThread->javaVM;
-
-		/* Release this thread (allow it to die) now that we are no longer inspecting it */
+#if JAVA_SPEC_VERSION >= 19
+	if (NULL != thread) {
+		j9object_t threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+		if ((NULL != threadObject) && IS_VIRTUAL_THREAD(currentThread, threadObject)) {
```

Comment:
If `currentThread` is a virtual thread, `getVMThread` will do early return without incrementing the `inspectorCount`.  The `releaseVMThread` logic here need to check for that case as well.

There is also the case that if the virtual thread is terminated, this code should not change the inspectorCount same as `getVMThread`. Though I am not sure such case should be handled here or by the caller as `getVMThread` will return an error code instead of the normal value.

Reply:
> releaseVMThread logic here need to check for that case as well.

We should add a `(threadObject != currentThread->threadObject)` check after this line or replace the `NULL != threadObject` check (is it necessary?)

> the case that if the virtual thread is terminated, this code should not change the inspectorCount same as getVMThread

If `mustBeAlive == true` and the thread is terminated then `releaseVMThread` will not be called (since error will be returned). If `mustBeAlive == false` we are handling platform threads correctly (check NULL != targetThread before modifying inspector count) but virtual threads do not have this check. I think we should add the check `vthreadState != JVMTI_VTHREAD_STATE_TERMINATED` (not `!isThreadAlive` since we want to prevent new threads from terminating as well)

Reply:
> We should add a (threadObject != currentThread->threadObject) check after this line or replace the NULL != threadObject check (is it necessary?)

You can just replace the NULL check on threadObject, since a valid jthread should never have a null threadObject associated.

`vthreadState != JVMTI_VTHREAD_STATE_TERMINATED` sounds good to me if the timing issue above in fixed.